### PR TITLE
Windows build performance

### PR DESCRIPTION
**Pull request by zodern**
### Before:
The times are from the todos app. Rebuilds are for editing a file that is only used in the client, but not in a `client` folder, causing both the server and clients to be rebuilt.

#### Initial build (second and later runs):
resolveContraints: 2,149 2,261
prepareProjectForBuild: 10,299 11,230
build app: 20,656 18,792
App process started after: 36,902 35,839
Creating watchers finished: 12,383 13,462ms after app started
AppProcess onListen: 23,900 25,644ms after app started

Usuable after: 60,802 61,483

#### Rebuild

prepareProjectForBuild: 6,713 6,674 7,158
rebuildApp: 4,555 4,269 4,008
App process started after: 11,690 11,499 11,621
Creating watchers finished: 4,536 4,057 4,001ms after app started
AppProcess onListen: 8,746 8,090 8,015ms after app started
Delayed build: 9,469 8,484 8,406ms after app started

Usuable after: 20,436 19,589 19,636
Delayed build at: 21,159 19,983 20,027

#### Initial build (second and later runs):
resolveContraints: 1,392 1,879 2,053
prepareProjectForBuild: 6,601 7,267 6,708
build app: 16,056 14,270 12,810
App process started after: 28,662 26,679 24,557
AppProcess onListen: 2,007 2,214 1,593 after app started
Watchers are created after onListen and are no longer blocking

Usuable after: 30,669 28,893 26,150

#### Rebuild

prepareProjectForBuild: 6,187 6,013 6,359
rebuildApp: 2,587 2,297 1,842
App process started after: 9,256 8,757 8,811
AppProcess onListen: 1,936 1,894 1,697 after app started
Delayed build: 6,255 6,522 6,551 after app started

Usuable after: 11,192 10,651 10,508
Usuable + delayed build: 15,511 15,279 15,362

### Notable Changes

The large app times are from the rebuild profile at https://github.com/meteor/meteor/issues/5091#issuecomment-444993610.

#### Delete garbage directories async
Saves 500 - 1,100ms per rebuild (large app 9s)

#### Do not use atomicallyRewriteFile when not building in place
Saves ~650ms (large app ~29,000ms including legacy bundle) per build.
This should be safe since the build is only used once all files have been successfully written. Renaming the two files takes almost twice as long as writing it.

#### isPortable
Saves ~900ms. Since npm-rebuilds.json is never used from the bundle created by "meteor run", creating it's contents is skipped.
The `.meteor-portable` file to cache the results was never created on Windows due to the file path being in the wrong format, which slowed down the initial build and `meteor build` when using local packages.

#### inPlace server builds on windows and caching _copyDirectory
Saves 1 - 1.5s per rebuild. Most of the warnings against doing this on windows give the reason of interfering with the running app. Since the server is always stopped during the build, it should be safe to update it in place.

If it is safe to update the server in place, it might also be safe to do the same with the legacy bundle since the server doesn't serve it while it is being written.

#### Create file watcher and start legacy bundle after server started
Both of those were blocking the event loop, causing a delay between the server starting, and the tool process logging '=> server started' and letting the proxy know to forward requests.

Watcher now has an option to do the initial check async. The delayed legacy bundle blocking has not been fixed.

### CODE REVIEW

**Comment by zodern** [comment]
After the last 4 comments, the build now is:
#### Initial
resolveConstraints: same
prepareProjectForBuild: 3,827 3,575 3,667
build app: 13,345 14,427 12,939
App process started after: 23,302 23,539 22,074
AppProcess onListen: same
Usuable after: 24,928

#### Rebuild client and server
prepareProjectForBuild: 1,762 797 958
rebuildApp: 2,333 1,817 1,723
App process started after: 4,485 2,915 2,999
Listened: same
Delayed build: 3,998 3,025 3,197ms after app started

Usable after 5,579
Usable + legacy: 9,577

#### Client only rebuild before last 4 commits
onChange: 2,562 3,389 3,136
prepareProjectForBuild: 4,444 5,029 3,542
rebuildApp: 1,580 1,438 1,180
Legacy built in 3,854 3,412 3,3842

Usable after 8,440

#### Client only rebuild after last 4 commits
onChange: 1,162 1,108 1,048
prepareProjectForBuild: 2,011 797 725
rebuildApp: 1,731 1,649 1,600
legacy built in: 1,365 1,129 1,144
Usable after 4,777 (first rebuild) 3,563 (subsequent rebuilds)

The client only rebuilds are timed before the server is rebuilt for the first time.

After the initial build, the watch sets seems to include every folder and some files in `node_modules`. The watch set created after a rebuild is missing those. This does make `onChange` very fast though, so causing the server to rebuild once after the app is started might make client rebuilds a couple seconds faster.

**Comment by CaptainN** [comment]
In my limited local testing, this PR makes rebuilds FASTER under Windows than under a VM on the same machine!

**Comment by CaptainN** [comment]
[fast-windows-meteor-rebuild.zip](https://github.com/meteor/meteor/files/2735369/fast-windows-meteor-rebuild.zip)
[fast-windows-meteor-startup.zip](https://github.com/meteor/meteor/files/2735370/fast-windows-meteor-startup.zip)

**Comment by yorrd** [comment]
Since this is not targeted at Linux, I didn't expect huge gains, but I can confirm that rebuild times are within variance and didn't change on Linux. I'd like to do some more accurate measurements but GC is getting in my way at the moment

**Comment by smeijer** [comment]
I'm having build times of more than 2 minutes. This patch would be more very, very welcome!

File: tools/isobuild/builder.js [hunk 1]
```diff
@@ -657,7 +680,9 @@ Previous builder: ${previousBuilder.outputPath}, this builder: ${outputPath}`
             // the read-only tools tree into a writable app."
             { mode: (fileStatus.mode & 0o100) ? 0o777 : 0o666 },
           );
+          }
```
Comment (hwillson) [inline tools/isobuild/builder.js]: Would you mind fixing the indentation of this block?

File: tools/runners/run-app.js [hunk 2]
```diff
@@ -835,7 +839,13 @@ _.extend(AppRunner.prototype, {
 
     Console.enableProgressDisplay(false);
 
-    const postStartupResult = runPostStartupCallbacks(bundleResult);
+    const postStartupResult = Promise.race([
+      listenPromise,
+      new Promise(resolve => setTimeout(resolve, 3000))
```
Comment (hwillson) [inline tools/runners/run-app.js]: The `3000` used here seems a bit arbitrary/magic-y. Maybe this should be defined in clearly labelled constant.

Comment (CaptainN) [reply tools/runners/run-app.js]: The commit message says "Also, the watcher and legacy bundle are only created after the server has started, or 3 seconds has passed." - it is necessary to have this 3 second time out at all? Why not always wait until the server starts?

**Comment by CaptainN** [comment]
Something is causing the spinner to remain after it is finished spinning:

```
=> Meteor server restarted                   -
I20190117-23:17:18.841(-5)? http://localhost:3000/
I20190117-23:17:18.863(-5)? http://localhost:3000/
I20190117-23:17:18.873(-5)? http://localhost:3000/
=> Meteor server restarted                   |
=> Meteor server restarted                   -
=> Meteor server restarted                   |
=> Meteor server restarted                   /
```

**Comment by CaptainN** [comment]
It seems that removing that 3 second race might have added back the lengthy end of build delay. I'll do more testing with and without and see if it reveals anything.

**Comment by smeijer** [comment]
Is this branch present in the current beta release?

**Comment by tabasco-addict** [comment]
To use the newest release, do i install normally using choco , then replace the files in the .meteor folder with the ones in the 1.8.1 repo?

**Comment by benjamn** [comment]
@tabasco-addict Install via choco (which will give you Meteor 1.8.0.2) and then run
```sh
meteor update --release 1.8.1-beta.16
```
in any application directory. You can also create a new app with
```sh
meteor create --release 1.8.1-beta.16 new-181b16-app
```

**Comment by smeijer** [comment]
Okay, this one is officially awesome! Thanks @zodern and @benjamn!

My build time (hot update) went down from 42 seconds to 9 seconds. Waiting for the browser refresh, takes another 5 seconds.

Usually, Meteor starts to be slower after a couple of hours development with a number of rebuilds. So I still need to monitor that one. But I have good hope.

I've installed Ubuntu 2 days ago. Turns out, that that was either a year too late, or a day to soon.

**Pull request merged**